feat: Improvements

Atinux opened the pull request:
Hey @jamespegg 

I am actually working on a better integration of Prismic with Nuxt.js

Here's my first work on it, looking forward for your review :)

- Removed `deferLoad` to have it by default + added at the end of `<body>`
- Use and prefers the `prismic` key in `nuxt.config.js` for module options
- Moved `linkResolver` key to its own file: `~/app/prismic/link-resolver.js`, same for `htmlSerializer` (this will become a standard with Nuxt 3)
- When `preview` mode on, will add `/preview` route + page
- Handles preview mode on statically generated websites :fire:
- Use the last version of preview mode (with `?repo=...` parameter)
- Add `<prismic-link>` and `<primic-image>` (based from https://github.com/prismicio/prismic-vue/tree/master/src/components but using functional components instead)
- Update documentation
- Update tests

Atinux commented:
If you want to discuss about it together, please ping me on Nuxt discord (https://discord.nuxtjs.org), my username is Atinux#1058

jamespegg commented on spec/index.spec.js:
```diff
@@ -14,59 +14,34 @@ describe("prismic-nuxt module", function() {
+  it("should add the components", function() {
     prismicNuxt.call(context, moduleOptions);
-    expect(context.options.head.script[1].src).toEqual('//static.cdn.prismic.io/prismic.min.js');
+    expect(context.addPlugin.mock.calls).toHaveLength(3)
```
This test is a little brittle because adding any more plugins would automatically break this test. Is there any way to check that the components have specifically been added? Maybe using `expect.stringContaining` or similar.

jamespegg commented on spec/index.spec.js:
```diff
@@ -14,59 +14,34 @@ describe("prismic-nuxt module", function() {
-    expect(context.options.head.script[1].defer).not.toBeDefined();
+  it("should not add the components if components=false", function() {
+    prismicNuxt.call(context, { ...moduleOptions, components: false });
+    expect(context.addPlugin.mock.calls).toHaveLength(1)
```
See above

jamespegg commented on docs/components.md:
```diff
@@ -0,0 +1,29 @@
+
+## &lt;prismic-link&gt;
+
+This component will depending of your prismic link use `<nuxt-link>` or `<a>`.
```
Just a small tweak :)

```suggestion
This component will, depending on your prismic link, use `<nuxt-link>` or `<a>`.
```

jamespegg commented on docs/components.md:
```diff
@@ -0,0 +1,29 @@
+
+## &lt;prismic-image&gt;
+
+This component will render an `<img>`, also completing `alt` and `copyright` attributes if provided from Prismic API.
```
```suggestion
This component will render an `<img>`, also completing `alt` and `copyright` attributes if provided by the Prismic API.
```

jamespegg commented on src/index.js:
```diff
@@ -2,35 +2,36 @@ const path = require('path');
-    this.options.head.__dangerouslyDisableSanitizersByTagID = {};
-  }
+  if (options.preview) {
+    const repo = options.endpoint.replace(/^https?:\/\//, '').replace(/(\.cdn)?\.prismic.+/, '');
```
Interesting, I'm wondering if might be easier if we just have the `repo` instead of the entire `endpoint`? Obviously a breaking change, but would make this a bit easier. The other positive is that we can control which version of the Prismic API this library uses.

Atinux replied:
I thought about it too, but actually it's harder to get the `repo` from Prismic UI than the endpoint :/

Atinux commented:
Hey @jamespegg 

I pushed my last changes, I moved the templates in `templates/` directory.
Also, with the next version of Nuxt (`v2.9.0`), users will be able to overwrite default templates provided by modules in the `app/` directory.

I also added a way to support the preview mode on generated websites.

You can see a demo of my PR here: https://github.com/Atinux/nuxt-prismic-showcase

About the tests, I will update them :)

jamespegg commented:
@Atinux just a few things and then I think we're good to go;

1. I'm not keen on forcing people to use a specific directory structure and filename for the `linkResolver` and `htmlSerializer`. Developers currently have the freedom of organising their code however they want and I wanted to avoid making the project too opinionated. I'm happy to include some updates in the documentation to suggest it as a best practice instead.

2. I'm not sure about moving the `preview` functionality into its own page, along with a logo and some styling. Can you explain your thoughts behind that?

3. Just need to get the unit test coverage up to 100% :)

As a general question, I've not quite figured out how to unit test template files which include `lodash` templates. Have you got any advice / examples around that by any chance?

Atinux commented:
Hey @jamespegg 

I will work to finish the PR today.

About 1., it's a new way we introduce for Nuxt users to extend the app options without forcing a full Nuxt reload (better developer experience), see https://nuxtjs.org/api/configuration-router/#scrollbehavior we introduced with Nuxt `v2.9.0`.

About 2., the idea is to avoid calling `$prismic.preview()` in every page and keep it only for a specific page, also, this page is not visible when using server-side rendering, but only for `spa` mode or using `nuxt generate`.

Working on 3. now :)

Atinux commented:
@jamespegg I updated the tests to have 100% coverage.

I also added a way to keep using `linkResolver` and `htmlSerializer` in the module options to keep retro-compatibility. I do believe this should be a major version to avoid any bad surprises and I think it could be nice to recommend to use Nuxt `>= v2.9.0`.

jamespegg commented:
@Atinux Okay, great. Yep, I think doing a major version bump makes sense and I'm happy to recommend Nuxt `>= 2.9.0` to users. I've had a couple of features on the go that I wanted in for `v1.0.0`, so I'll finish those up and put everything in together.

Atinux commented:
Awesome thank you 💚

hypervillain commented:
👋Hey guys,

I think we could take this PR as an opportunity to ditch the [prismic-vue Nuxt branch](https://github.com/prismicio/prismic-vue/tree/nuxt) in favour of `prismic-nuxt`. I looked at the repository and feel like part of @Atinux's code could benefit to `prismic-vue` and be imported by `prismic-nuxt`. How I see it:

- Functional components (eg. PrismicImage) should belong to `prismic-vue` so they can be used in non-Nuxt projects too. It's a great improvement that everyone could benefit from. If I understand correctly, only `PrismicLink` here is Nuxt specific, so we could whether create a `PrismicNuxtLink` in `prismic-vue` or have it here - whichever you feel is the most natural.

- We could also import the `RichText` component from `prismic-vue`. In terms of developer experience, a lot of people are used to `<prismic-rich-text :field="field" />` and keeping it that way would facilitate the migration of existing Prismic-Nuxt projects. I can make a PR to rewrite the component and make it functional of course (plus the others) to `prismic-vue`.

- Add a migration guide somewhere on Prismic documentation and/or in `/docs` so that existing Prismic users don't get lost. I'll make sure to deprecate the Nuxt branch of `prismic-vue` in the meantime.

- This is one minor but I'm not fond of `Prismic.Predicates` becoming `$prismic.predicates`: when one of our developers tried to migrate they got stuck after updating their requests. Maybe we could make one the alias of the other?

I'd love to have your opinion on this,
Hugo

Atinux commented:
I do agree that `prismic-vue` and `prismic-nuxt` should get along and make the transition from a Vue project to Nuxt as seamless as possible 👍 

The advantage of `Prismic.Predicates` becoming `$prismic.predicates` is the ability to avoid importing `Prismic` in the pages/components (and it is already an alias :))

hypervillain commented:
Perfect,

I started working on `prismic-vue` this morning and will share a link to the PR right here 👍

I thought more of having it this way: `$prismic.Predicates` but it may not be the most appropriate solution - not a big deal anyway!

hypervillain commented:
Hey guys,

here is my PR on `prismic-vue`: https://github.com/prismicio/prismic-vue/pull/14
I'll need to try and use it with an actual Nuxt project but it works with our Vue starter kits!

Components are now functional and are bundled separately.
Basically, the changes needed in `prismic-nuxt` would be something like:

```javascript
import { common, nuxt } from 'prismic-vue/components';

if (options.components) {
  Object.entries({ ...common, ...nuxt }).forEach(([_, c]) => {
    Vue.component(c.name, c)
  })
}
```

Let me know if this works for you of if you need some changes on `prismic-vue` side

Atinux commented:
Nice work @hypervillain 

It sounds good to me, waiting for you to release the v2 of `prismic-vue` before updating this PR :)

I would love to have @jamespegg thoughts on this

jamespegg commented:
Hi @hypervillain @Atinux,

I've been mulling over this for a few days, and here are my thoughts;

Firstly, from a technical perspective I've got no issue around where components sit, and I agree that because they're useful more generally as Vue components that they belong in `prismic-vue`. The current released version (`v0.5.3`) doesn't itself ship with any components, so now is the perfect time to decide where these go. Currently there is only one thing in the backlog (#21) that would require a component, but I was thinking about splitting this out anyway, and it may even belong in `prismic-vue` if appropriate.

This sort of sits better with the initial intention of `prismic-nuxt`, which was to be a small lightweight library that just makes it easy to get start with Prismic and Nuxt. Adding more and more to it out of the box has the potential of making it unwieldy.

I'm also thinking this applies for other things that I've got on the go, like auto generating sitemaps (#20). It should really be an optional thing that developers can include, so might belong in a separate but related module.

Secondly, just a request from my end is that we split future PR's down into smaller chunks so that I can review them quicker and get them through. It feels like we're heading for a good `v1.0.0` release, but I still have to be able to get security patches through in a timely manner. So I'm going to adopt gitflow and edit this PR to go into `develop`, and ask that any future PR's also go into `develop`. If you need to get any urgent hotfixes in, then obviously put a PR straight into `master`.

Finally, I need to have a think about what this means for the project more broadly. It's currently a very fledgling project, with only 200 weekly downloads on NPM. With `prismic-vue` being an official Prismic plugin, and `prismic-nuxt` being fully community driven... I'm not totally clear how that plays out. Maybe a good first step would be to add you guys as collaborators so that you can at least complete peer reviews and merge things into `develop`?

Cheers,
James

Atinux commented:
Hi @jamespegg 

I agree on the fact that components should live in `prismic-vue` (as well as the [slice component](https://github.com/jamespegg/prismic-nuxt/issues/21)).

About [sitemap](https://github.com/jamespegg/prismic-nuxt/issues/20), I would recommend people to use the nuxt sitemap module (and maybe add a note in prismic-nuxt doc to show an example): https://github.com/nuxt-community/sitemap-module

100% agree with `develop` branch.

> Maybe a good first step would be to add you guys as collaborators so that you can at least complete peer reviews and merge things into develop?
Sure :)

Atinux commented:
Hey @jamespegg,

Any news about when this could be merged? 🙏

hypervillain commented:
Hey guys, sorry for my poor response time these last weeks..

prismic-vue V2 should be published today tomorrow, so I think we could merge the PR as soon as it's integrated.

About #21, I think I have something that might be of interest for you. It's still pretty raw, undocumented but works with another project I'm working on. I'd love you guys to have a look at it: https://gist.github.com/hypervillain/d5b8c400f324a6313874e8d3927e1f97 - I'll make it a bit more generic/ less bloated regarding dynamic import paths but I'd be happy to release it soon!

What I would love to add to the plugin is an auto-generate command for Prismic related pages. I've tried and failed to do it 100% automatically but I'm sure that with the right amount of configuration, everyone would gain a lot of time from this..

> With prismic-vue being an official Prismic plugin, and prismic-nuxt being fully community driven... I'm not totally clear how that plays out.

I don't know exactly either. I think that moving the next stable version to Nuxt community is something to consider as it would give users a clearer understanding that this is the package that they should use. As of today, people still ask us why we don't have an official package, so let's make things official :)

Would it work for you @jamespegg?
@Atinux, is adding this module to Nuxt community something that you would consider?

Atinux commented:
I am ok with that, as well as having @nuxtjs/prismic for namespace in order to make it more official on the community space of Nuxt.

For the Slice component, from what I see it looks nice, only thing is that the path is not super sexy (vueSlices/slices), but we can change the default when used with Nuxt I guess (I am thinking of @/app/prismic/slices).

jamespegg commented:
@Atinux @hypervillain Sorry guys, things have been crazy busy the last few weeks.

Yep, let's move this to under Nuxt. Unfortunately I haven't really got the time at the moment to look after the library and try and grow it on my own, so moving it over makes sense. I'll still try and contribute as and when I can. :)

How do we want to handle this? There's three areas we need to cover;

1. **The repository itself**. I'm happy to transfer it to the Nuxt organisation. I'm equally happy for @Atinux to fork it for the Nuxt org and for me to mark this one as deprecated / archived and to link to the new Nuxt one.

2. **The `prismic-nuxt` npm package**. I guess the only option here is to deprecate it? I guess moving it another org and changing it to `@nuxt/prismic` would be a breaking change.

3. **The docs site (prismic-nuxt.js.org)**. This feeds directly off GitHub pages, so I'm not directly managing any infrastructure for it. Happy for it to carry on as is, redirect to official Nuxt documentation or redirect to official Prismic documentation. It's only getting 100 monthly users, so the impact is going to be fairly minimal.

jamespegg commented:
Also as an extra thing, I've just added both of you as collaborators to make things easier.

Atinux commented:
Hi @jamespegg 

Thank you very much for your answer, I could also transfer it directly to `nuxt-community` instead of forking it if it's ok for you? I will add the credits that you are the original author of this package <3

Then publish the `@nuxtjs/prismic` package while deprecating the `prismic-nuxt` package?

For the docs, we can keep it on the same domain if it's ok? I think a PR to https://github.com/js-org/js.org/blob/master/cnames_active.js#L1391 will be enough :)

jamespegg commented:
Hi @Atinux ,

Perfect, happy for you to transfer instead of fork, keeps it cleaner. If you let me know once you’ve published the new package, I’ll deprecate the old one. Similarly with the PR for js.org, I’ll comment on it and say it’s okay just in case they have questions.

Do you have a rough idea of when you want to move everything? I’ll try and make myself available to help if needed.

Atinux commented:
@jamespegg could you make me admin of this repo please so I can make the transfert on my own without bothering you :)

As weel as publish right to https://www.npmjs.com/package/prismic-nuxt (my username is `atinux`) so I can deprecate it.

Thank you!

jamespegg commented:
@Atinux  I've just added you to the NPM repo, so you should be good to deprecate that. I can't seem to find the option to make you an admin of the GitHub repo, so I might just need to transfer ownership to you?

Atinux commented:
@jamespegg well if you can transfer the ownership that would be nice indeed.

Atinux commented:
Sorry to bother @jamespegg but I would like to finish all these steps before new year, could you give me ownership to the prismic-nuxt repo please?

jamespegg commented:
@Atinux You need to delete your fork first, then I can transfer it

Atinux commented:
Forked removed @jamespegg

jamespegg commented:
@Atinux Okay, transfer request all done :)
